sharding/shard.go

```diff
@@ -0,0 +1,219 @@
+// shardID/period pair
+func (s *Shard) CanonicalCollationHash(shardID *big.Int, period *big.Int) (*common.Hash, error) {
+	key := canonicalCollationLookupKey(shardID, period)
+	hash := common.BytesToHash(key.Bytes())
+	collationHashBytes, err := s.shardDB.Get(hash)
+	if err != nil {
+		return nil, fmt.Errorf("no canonical collation set for period, shardID pair: %v", err)
+	}
+	collationHash := common.BytesToHash(collationHashBytes)
+	return &collationHash, nil
+}
+
+// CanonicalCollation fetches the collation set as canonical in the shardDB.
+func (s *Shard) CanonicalCollation(shardID *big.Int, period *big.Int) (*Collation, error) {
+	h, err := s.CanonicalCollationHash(shardID, period)
+	if err != nil {
+		return nil, fmt.Errorf("hash not found: %v", err)
+	}
+	collation, err := s.CollationByHash(h)
+	if err != nil {
+		return nil, fmt.Errorf("no canonical collation found for hash: %v", err)
+	}
+	return collation, nil
+}
+
+// BodyByChunkRoot fetches a collation body.
+func (s *Shard) BodyByChunkRoot(chunkRoot *common.Hash) ([]byte, error) {
+	body, err := s.shardDB.Get(*chunkRoot)
+	if err != nil {
+		return nil, fmt.Errorf("no corresponding body with chunk root found: %v", err)
+	}
+	return body, nil
+}
+
+// CheckAvailability is used by notaries to confirm a header's data availability.
+func (s *Shard) CheckAvailability(header *CollationHeader) (bool, error) {
+	key := dataAvailabilityLookupKey(header.ChunkRoot())
+	availabilityVal, err := s.shardDB.Get(key)
+	if err != nil {
+		return false, fmt.Errorf("key not found: %v", key)
+	}
+	var availability int
+	if err := rlp.DecodeBytes(availabilityVal, &availability); err != nil {
+		return false, fmt.Errorf("cannot RLP decode availability: %v", err)
+	}
+	if availability != 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+// SetAvailability saves the availability of the chunk root in the shardDB.
+func (s *Shard) SetAvailability(chunkRoot *common.Hash, availability bool) error {
+	key := dataAvailabilityLookupKey(chunkRoot)
+	if availability {
+		enc, err := rlp.EncodeToBytes(true)
+		if err != nil {
+			return fmt.Errorf("cannot RLP encode availability: %v", err)
+		}
+		s.shardDB.Put(key, enc)
+	} else {
+		enc, err := rlp.EncodeToBytes(false)
+		if err != nil {
+			return fmt.Errorf("cannot RLP encode availability: %v", err)
+		}
+		s.shardDB.Put(key, enc)
+	}
+	return nil
+}
+
+// SaveHeader adds the collation header to shardDB.
+func (s *Shard) SaveHeader(header *CollationHeader) error {
+	encoded, err := header.EncodeRLP()
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %v", err)
+	}
+
+	// Uses the hash of the header as the key.
+	hash := header.Hash()
+	s.shardDB.Put(hash, encoded)
+	return nil
+}
+
+// SaveBody adds the collation body to the shardDB and sets availability.
+func (s *Shard) SaveBody(body []byte) error {
+	// TODO: check if body is empty and throw error.
+	// TODO: dependent on blob serialization.
```

Comment:
For these TODOs, can you add github issues? People are hungry to contribute and these would be easy tasks